docs/src/proposals/comprehensive-compute-fees.md

```diff
@@ -0,0 +1,88 @@
+---
+title: Comprehensive Compute Fees
+---
+
+## Motivation
+
+The current fee structure lacks a comprehensive account of the work required by
+a validator to process a transaction.  The fee structure is only based on the
+number of signatures in a transaction but is meant to account for the work that
+the validator must perform to validate each transaction.  The validator performs
+a lot more user-defined work than just signature verification.  The work to
+process a transaction typically includes signature verifications, account
+locking, account loading, and instruction processing.
+
+## Proposed Solution
+
+### New fee structure
+
+In addition to signature checking, the other work required to process an
+instruction should be taken into account.  To do this, the total fee should be
+an accumulated cost of all the work.  Each piece of work can be
+measured/accounted for in a way that makes sense for the type of work involved.
+
+A new fee structure could include:
+1. flat fee per signature
+2. flat fee for each write lock
+3. per-byte fee for the amount of data in each loaded account
+4. per-compute-unit fee for the number of compute units used to process an
+   instruction
+
+Fees 1-3 can be determined upfront before the message is processed.  #4 would
+need to be finally accounted for after the message is processed, but the payer's
+balance could be pre-deducted against a compute budget cap to ensure they have
+enough lamports to cover a max fee, and then any unused compute budget could be
+credited back after the message is fully processed.
+
+The goal of the fees is to cover the computation cost of processing a
+transaction.  Each of the above fee categories could be represented as a compute
+unit cost that, when added together, encompasses the entire cost of processing
+the transaction.  By calculating the total cost of the transaction the runtime
+can make better decisions on what transactions to process and when, as well as
+throw out transactions that exceed a cap.
+
+The per-compute unit fee doesn't have to be linear; developers could be
+non-linearly incentivized to reduce compute costs by optimizing in any of the 4
+fee categories listed above.
+
+To give developers some control over fees and the compute cap, a new built-in
+instruction could be introduced that requests a specific transaction-wide
+compute budget cap.  The instruction can be used by a developer to reduce the
+cap and thus fees they expect to pay, or to request a higher than the default
+cap.  The runtime could in-turn, use these instructions to determine how to
+schedule these more expensive transactions.
+
+### Transaction compute caps
```

Comment:
@sakridge, @mvines @jstarry  Planning to start with applying a transaction-wide cap, thoughts?

Reply:
(not sure I have much to add here at the moment)

Reply:
I'm looking for opinions on how much of an effect this would have on developers and if applying a tx wide cap makes sense where each instruction charges against a common meter (rather then what we have now where each instruction has its own budget)

Reply:
I mean I think we need to do it.  What impact it has could be determined by running a hacked validator against mainnet to see how many existing transactions would start failing when we do this.

Reply:
+1 impact would depend on what transactions are going into the system now and then old vs. new costs.